### PR TITLE
Support for testing without provisioner converging

### PR DESCRIPTION
**Pull request by jbartko**
__User Story:__
As a docker microservice operator, I would like to leverage kitchen-dokken to
perform integration testing against arbitrary containers.

__Test Case:__
Given: the Test Kitchen provisioner can be fooled into not converging
Given: verifier tests such as Inspec can be used to validate containers under
test without a provisioner having run
When: testing containers without first converging
Expect: to be able to set the container environment
Expect: to be able to set the container entry point

### CODE REVIEW

**Comment by someara** [comment]
Hello!

I'll have some time to give this a proper review some time next week. Thanks for the PR!

-s

File: .kitchen.yml [hunk 1]
```diff
@@ -15,37 +15,51 @@ verifier:
...
-  driver:
-    image: fedora:latest
-    pid_one_command: /usr/lib/systemd/systemd
-    intermediate_instructions:
-      - RUN dnf -y install yum which net-tools procps-ng telnet iputils nmap
-  run_list:
-    - recipe[chef-yum-docker]
+  - name: hello
+    driver:
+      image: alpine:latest
```
Comment (jbartko) [inline .kitchen.yml]: Example "arbitrary" container -- this is representative of an  in-house "fooapp" container.

File: .kitchen.yml [hunk 2]
```diff
@@ -15,37 +15,51 @@ verifier:
...
+      env:
+        - FOO=BAR
```
Comment (jbartko) [inline .kitchen.yml]: Container behaviour is often driven by environment. Test case requirement.

File: .kitchen.yml [hunk 3]
```diff
@@ -15,37 +15,51 @@ verifier:
...
+    provisioner:
+      chef_binary: /bin/true
```
Comment (jbartko) [inline .kitchen.yml]: A no-op binary to "fool" the provisioner. Test case assumption.

File: .kitchen.yml [hunk 4]
```diff
@@ -15,37 +15,51 @@ verifier:
...
-    driver:
-      chef_version: latest
+    excludes:
+      - hello
     run_list:
-    - recipe[dokken_test::default]
+      - recipe[dokken_test::default]
+    attributes:
+      dokken_test:
+        revision: <%= `git rev-parse HEAD` %>
```
Comment (jbartko) [inline .kitchen.yml]: Expose the sha from the current commit/PR to the Fedora container under test.

File: .kitchen.yml [hunk 5]
```diff
@@ -15,37 +15,51 @@ verifier:
...
 
   - name: hello
     driver:
-      chef_version: latest
+      entrypoint:
+        - /bin/sh
+        - -c
+        - while :; do nc -lvp 1234 -e /bin/sh -c 'echo hello'; done
```
Comment (jbartko) [inline .kitchen.yml]: Replaces `dokken_test::hello` recipe.

File: .kitchen.yml [hunk 6]
```diff
@@ -15,37 +15,51 @@ verifier:
...
+      ports: '1234'
+      pid_one_command: ''
+    excludes:
+      - fedora-latest
 
   - name: helloagain
     driver:
-      chef_version: latest
       hostname: helloagain.computers.biz
+      intermediate_instructions:
```
Comment (jbartko) [inline .kitchen.yml]: Move example intermediate_instructions from Fedora container under test to the 'helloagain' alpine container under test.

File: lib/kitchen/driver/dokken.rb [hunk 7]
```diff
@@ -149,10 +151,14 @@ def build_work_image(state)
       end
 
       def work_image_dockerfile
-        from = "FROM #{platform_image}"
-        custom = ['RUN /bin/sh -c "echo Built with Test Kitchen"']
-        Array(config[:intermediate_instructions]).each { |c| custom << c }
-        [from, custom].join("\n")
+        dockerfile_contents = [
+          "FROM #{platform_image}",
+          'LABEL X-Built-By=kitchen-dokken'
```
Comment (jbartko) [inline lib/kitchen/driver/dokken.rb]: Should not assume `/bin/sh` availability. Replace with a `LABEL` Dockerfile directive.

File: lib/kitchen/driver/dokken.rb [hunk 8]
```diff
@@ -270,6 +277,9 @@ def start_runner_container(state)
             },
           },
         }
+        unless self[:entrypoint].to_s.length == 0
```
Comment (jbartko) [inline lib/kitchen/driver/dokken.rb]: Avoid overriding the Entrypoint to `nil` if unspecified.

File: test/cookbooks/dokken_test/recipes/default.rb [hunk 9]
```diff
@@ -1,23 +1,38 @@
...
+docker_service 'default' do
+  host ['tcp://127.0.0.1']
+  action [:create, :start]
 end
 
 git '/home/notroot/kitchen-dokken' do
-  repository 'https://github.com/someara/kitchen-dokken'
-  revision 'master'
+  repository '/opt/kitchen-dokken/.git'
+  revision node['dokken_test']['revision']
```
Comment (jbartko) [inline test/cookbooks/dokken_test/recipes/default.rb]: Needed a way for Travis to build the gem from the PR commit instead of from the master branch.

**Comment by jbartko** [comment]
@someara,

I've sprinkled a few rationale comments into the PR. My organization is actively using these changes to perform integration testing against in-house containerized applications with public container dependencies.

**Comment by jbartko** [comment]
@someara,

I've rebased this PR against master and made a few other tweaks to get the Travis build passing again.

**Pull request merged**